File: simplecrypto/simplecrypto.js
// This file implements a similar interface as the Go library, but in node.js.
// This is only for demonstration, and isn't intended to be imported directly
// with something like npm without some work.

var crypto = require('crypto');

/**
 * Generates an algorithm name based on a key's size. I.e., providing a
 * 16 byte key will return "aes-128-cfb".
 * @param {Buffer} key The AES key used for encryption.
 * @return {string} The algorithm name to give to createCipher.
 */
function algoFromKey(key) {
    return 'aes-' + (key.length * 8) + '-cfb';
}

/**
 * Encrypts a payload with AES CFB.
 * @param {Buffer} key The AES key used for encryption.
 * @param {Buffer} payload The payload to be encrypted.
 * @return {Buffer} Ciphertext (IV + encrypted payload).
 */
function encrypt(key, payload) {
    const algo = algoFromKey(key);

    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv(algo, key, iv);

    let encrypted = cipher.update(payload);
    encrypted = Buffer.concat([encrypted, cipher.final()]);

    return Buffer.concat([iv, encrypted]);
}

/**
 * Decrypts a payload with AES CFB.
 * @param {Buffer} key The AES key used for decryption.
 * @param {Buffer} ciphertext Ciphertext (IV + encrypted payload).
 * @return {Buffer} The original payload. 
 */
function decrypt(key, ciphertext) {
    const algo = algoFromKey(key);
    
    const iv = ciphertext.slice(0, 16);
    ciphertext = ciphertext.slice(16);

    const decipher = crypto.createDecipheriv(algo, key, iv);

    let payload = decipher.update(ciphertext);
    payload = Buffer.concat([payload, decipher.final()]);
    return payload;
}


/**
 * Calculates the HMAC of the message using SHA256.
 * @param {Buffer} key The encryption key.
 * @param {Buffer} message The message being HMAC'd.
 * @return {Buffer} The calculated MAC.
 */
function hmac(key, message) {
    const mac = crypto.createHmac('sha256');
    mac.update(message);
    return mac.digest();
}

/**
 * Checks whether messageMAC is a valid HMAC tag for a message.
 * @param {Buffer} key The encryption key.
 * @param {Buffer} message The message being HMAC'd.
 * @param {Buffer} messageMAC The (proposed) MAC of the message.
 * @return {boolean} True if the MACs match.
 */
function checkMAC(key, message, messageMAC) {
    const expectedMAC = hmac(key, message);
    return crypto.timingSafeEqual(messageMAC, expectedMAC);
}

File: simplecrypto/simplecrypto.go
// Package simplecrypto implements a simplified interface for encryption and
// decryption using AES encryption in CFB mode.
//
// It is implemented using examples given in crypto/cipher and crypto/hmac.
package simplecrypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"io"
)

// ErrCiphertextTooShort is returned by the Decrypt function when the provided
// ciphertext is too short to contain a valid IV.
var ErrCiphertextTooShort = errors.New("simplecrypto: ciphertext too short")

// Encrypt encrypts a payload using the given key. It returns a byte
// slice with the IV as the first aes.BlockSize bytes, followed by
// the encrypted payload.
//
// Note: Ciphertexts must be authenticated as well as encrypted in order to
// be secure. Be sure to calculate the ciphertext's HMAC to send with it.
// This library provides shorthand for calculating the HMAC of a ciphertext.
func Encrypt(key, payload []byte) (ciphertext []byte, err error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	ciphertext = make([]byte, aes.BlockSize+len(payload))

	iv := ciphertext[:aes.BlockSize]
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, err
	}

	stream := cipher.NewCFBEncrypter(block, iv)
	stream.XORKeyStream(ciphertext[aes.BlockSize:], payload)

	return ciphertext, nil
}

// Decrypt decrypts a ciphertext using the given key. It returns a byte
// slice of the encrypted payload.
//
// This function assumes the ciphertext is in the format generated by the
// Encrypt function, i.e. the IV followed by the encrypted payload.
//
// Note: Ciphertexts must be authenticated as well as encrypted in order to
// be secure. Be sure to check the ciphertext's HMAC before decrypting it.
// This library provides shorthand for checking the HMAC of a ciphertext.
func Decrypt(key, ciphertext []byte) (payload []byte, err error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	if len(ciphertext) < aes.BlockSize {
		return nil, ErrCiphertextTooShort
	}

	iv := ciphertext[:aes.BlockSize]
	ciphertext = ciphertext[aes.BlockSize:]

	stream := cipher.NewCFBDecrypter(block, iv)
	stream.XORKeyStream(ciphertext, ciphertext)

	return ciphertext, nil
}

// HMAC calculates the HMAC of the message using the given key using the
// SHA256 has function.
func HMAC(key, message []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write(message)
	return mac.Sum(nil)
}

// CheckMAC reports whether messageMAC is a valid HMAC tag for message.
//
// This implementation is given in the documentation for crypto/hmac.
func CheckMAC(key, message, messageMAC []byte) bool {
	expectedMAC := HMAC(key, message)
	return hmac.Equal(messageMAC, expectedMAC)
}
